dev/bots/codesign.dart

```diff
@@ -0,0 +1,105 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:io';
+import 'package:path/path.dart' as path;
+
+String get repoRoot => path.normalize(path.join(path.dirname(Platform.script.toFilePath()), '..', '..'));
+String get cacheDirectory => path.normalize(path.join(repoRoot, 'bin', 'cache'));
+
+/// Check mime-type of file at [filePath] to determine if it is binary
+bool isBinary(String filePath) {
+  final ProcessResult result = Process.runSync(
+    'file',
+    <String>[
+      '--mime-type',
+      '-b', // is binary
+      filePath,
+    ],
+  );
+  return (result.stdout as String).contains('application/x-mach-binary');
+}
+
+/// Find every binary file in the given [rootDirectory]
+List<String> findBinaryPaths([String rootDirectory]) {
+  rootDirectory ??= cacheDirectory;
+  final ProcessResult result = Process.runSync(
+    'find',
+    <String>[
+      rootDirectory,
+      '-type',
+      'f',
+      '-perm',
+      '+111', // is executable
+    ],
+  );
+  final List<String> allFiles = (result.stdout as String).split('\n').where((String s) => s.isNotEmpty).toList();
+  return allFiles.where(isBinary).toList();
+}
+
+/// Given the path to a stamp file, read the contents.
+///
+/// Will throw if the file doesn't exist.
+String readStamp(String filePath) {
+  final File file = File(filePath);
+  if (!file.existsSync()) {
+    throw 'Error! Stamp file $filePath does not exist!';
+  }
+  return file.readAsStringSync().trim();
+}
+
+/// Return whether or not the flutter cache is up to date.
+bool checkCacheIsCurrent() {
+  try {
+    final String dartSdkStamp = readStamp(path.join(cacheDirectory, 'engine-dart-sdk.stamp'));
+    final String engineVersion = readStamp(path.join(repoRoot, 'bin', 'internal', 'engine.version'));
+    return dartSdkStamp == engineVersion;
+  } catch (e) {
+    print(e);
+    return false;
+  }
+}
+
+void main() {
+  final List<String> failures = <String>[];
+
+  if (!Platform.isMacOS) {
+    print('Error! Expected operating system "macos", actual operating system '
+      'is: "${Platform.operatingSystem}"');
+    exit(1);
+  }
+
+  if (!checkCacheIsCurrent()) {
+    print(
+      'Warning! Your cache is either not present or not matching your flutter\n'
+      'version. Run a `flutter` command to update your cache, and re-try this\n'
+      'test.');
+    exit(1);
+  }
+
```

Comment:
Should we run flutter precache --no-android to make sure the iOS engine artifacts are actually present?

Reply:
Good idea, I created [this issue](https://github.com/flutter/flutter/issues/50384) to track. For the sake of landing this hotfix, I won't cherry pick it here. This test will still catch when the batch codesigning job hasn't been run, as the dart sdk, libimobiledevice artifacts, and android gen snapshots are all cached by `flutter doctor -v` run in the setup_script (see https://cirrus-ci.com/task/6198934779461632).